python/ray/_private/conda.py

```diff
@@ -52,3 +55,135 @@ def get_conda_bin_executable(executable_name):
+
+
+def _get_conda_env_name(conda_env_path):
+    conda_env_contents = open(conda_env_path).read() if conda_env_path else ""
+    return "ray-%s" % hashlib.sha1(
+        conda_env_contents.encode("utf-8")).hexdigest()
+
+
+def get_or_create_conda_env(conda_env_path, base_dir=None):
+    """
+    Given a conda YAML, creates a conda environment containing the required
+    dependencies if such a conda environment doesn't already exist. Returns the
+    name of the conda environment, which is based on a hash of the YAML.
+
+    Args:
+        conda_env_path: Path to a conda environment YAML file.
+        base_dir (str, optional): Directory to install the environment into via
+            the --prefix option to conda create.
+    Returns:
+        The name of the env, or the path to the env if base_dir is specified.
+            In either case, the return value should be valid to pass in to
+            `conda activate`.
+    """
+    conda_path = get_conda_bin_executable("conda")
+    try:
+        exec_cmd([conda_path, "--help"], throw_on_error=False)
+    except EnvironmentError:
+        raise ValueError(
+            "Could not find Conda executable at {0}. "
+            "Ensure Conda is installed as per the instructions at "
+            "https://conda.io/projects/conda/en/latest/"
+            "user-guide/install/index.html. "
+            "You can also configure Ray to look for a specific "
+            "Conda executable by setting the {1} environment variable "
+            "to the path of the Conda executable".format(
+                conda_path, RAY_CONDA_HOME))
+    (_, stdout, _) = exec_cmd([conda_path, "env", "list", "--json"])
+    envs = json.loads(stdout)["envs"]
+
+    project_env_name = _get_conda_env_name(conda_env_path)
+    if base_dir:
+        project_env_name = f"{base_dir}/{project_env_name}"
+        if project_env_name not in envs:
+            logger.info("=== Creating conda environment %s ===",
+                        project_env_name)
+            exec_cmd(
+                [
+                    conda_path, "env", "create", "--file", conda_env_path,
+                    "--prefix", project_env_name
+                ],
+                stream_output=True)
+        return project_env_name
+    else:
+        env_names = [os.path.basename(env) for env in envs]
+        if project_env_name not in env_names:
+            logger.info("=== Creating conda environment %s ===",
+                        project_env_name)
+            exec_cmd(
+                [
+                    conda_path, "env", "create", "-n", project_env_name,
+                    "--file", conda_env_path
+                ],
+                stream_output=True)
+        return project_env_name
+
+
+class ShellCommandException(Exception):
+    pass
+
+
+def exec_cmd(cmd,
+             throw_on_error=True,
+             env=None,
+             stream_output=False,
+             cwd=None,
+             cmd_stdin=None,
+             **kwargs):
```

Comment:
I believe the only option we need here is `stream_output` so we can simplify this a lot